assemble-and-restart

```diff
@@ -2,6 +2,32 @@
 set -x
 set -eo pipefail
 
+# If WorkingDir is injected as an env from odo and destination path is not equal to WorkingDir injected by odo
+if [ ! -z "${ODO_S2I_WORKING_DIR}" ] && [ "${ODO_S2I_SRC_BIN_PATH}" != "${ODO_S2I_WORKING_DIR}" ]; then
+    # If ODO_SRC_BACKUP_DIR is injected by odo
+    if [ -n  "$ODO_SRC_BACKUP_DIR" ]; then
+        # If it doesn't exit create it
+        if [ ! -d ${ODO_SRC_BACKUP_DIR} ]; then
+            mkdir -p ${ODO_SRC_BACKUP_DIR}/src
+        fi
+        # Backup the sources in DestinationDir to ODO_SRC_BACKUP_DIR bcoz the assemble script for some s2i images
+        # moves sources from DestinationDir to WorkingDir thereby deleting it there(DestinationDir). So, back it up
+        # for partial push use cases like watch
+        rsync -r ${ODO_S2I_SRC_BIN_PATH}/src/. ${ODO_SRC_BACKUP_DIR}/src/
+    fi
+    # Clear all those dirs in WorkingDir that exist in DestinationDir so that when assemble script moves sources from
+    # DestinationDir to WorkingDir, it sees the directory clean and doesn't complain:
+    # https://github.com/redhat-developer/odo/issues/1054
+    for file in `ls -a ${ODO_S2I_SRC_BIN_PATH}/src/`
+    do
+        if [ $file != "." ] && [ $file != ".." ]; then
+            rm -fr $ODO_S2I_WORKING_DIR/$file
+        fi
+    done
+    # In case of java clear off the target directory
+    rm -fr ${ODO_S2I_SRC_BIN_PATH}/src/target
```

Comment:
This is needed because java s2i does not delete the previous compilation output. If this weren't added, then deleted source files would not be reflected in the compilation output

Reply:
Unfortunately, if we don't delete this directory, watch + delete is totally broken for Java.
Perhaps we should wrap the delete call in an if statement that checks if the image is a Java image (by testing for the existence of the mvn binary)?

Reply:
Yeah I added the conditional check now
@kadel @geoand can you have a look now if it is ok?

Reply:
@anmolbabu I don't think that using `"$ODO_S2I_BUILDER_IMG" == "redhat-openjdk-18/openjdk18-openshift"` is correct, because the image builder could have a different name, right?
I personally think that testing for the image capabilities (meaning testing if it has the `mvn` binary) is the correct approach

Reply:
@geoand IMO, this fix is very specific to the way openjdk18 image redhat-openjdk-18/openjdk18-openshift works but java has ways to mitigate it like for example using `mvn clean package` which any other java image might be handling so, we need to check this image in particular?

Reply:
The only other Java builder we use at the moment is the widlfly one and I haven't checked to see if it uses `mvn clean`.
My issue is basically with the name - are you sure the builder image name that is injected as an env var will always be the same no matter how I name it in my cluster?

Reply:
Yeah this is the name of docker image corresponding to s2i image...
So it is likely not to change

Reply:
In that case, I'm fine with the change as long as we check and see if widlfly needs the same kind of handling :)

Reply:
What is exactly the issue that this fixes? I have tried a few Java examples and with this commented out and haven't any issues. Except hitting  https://github.com/redhat-developer/odo/issues/1117 again.

Reply:
@kadel To see the problem this fixes you need to do the following:

Use the updated source example from this PR.

```
odo create ... 
odo push
odo watch
```

Then delete one of the Java classes. Once the watch picks up the delete, the build should fail (since of the required sources no longer exists).
If you remove this line, then the build does not fail because it retains the previous compiled file